plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c

```diff
@@ -32,8 +37,104 @@ struct FuPluginData {
 	gboolean       needs_forcepower;
 	gboolean       updating;
 	guint          timeout_id;
+	gboolean       bolt_supported;
+	gint           bolt_fd;
 };
 
+static void
+fu_plugin_thunderbolt_power_bolt_supported (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_autoptr(GDBusConnection) connection = NULL;
+	g_autoptr(GVariant) val = NULL;
+	g_autoptr(GDBusProxy) proxy = NULL;
+	g_autoptr(GError) error_local = NULL;
+
+	connection = g_bus_get_sync (G_BUS_TYPE_SYSTEM, NULL, &error_local);
+	if (connection == NULL) {
+		g_warning ("Failed to initialize d-bus connection: %s",
+			   error_local->message);
+		return;
+	}
+
+	proxy = g_dbus_proxy_new_sync (connection,
+					G_DBUS_PROXY_FLAGS_NONE,
+					NULL,
+					BOLT_DBUS_SERVICE,
+					BOLT_DBUS_PATH,
+					BOLT_DBUS_INTERFACE,
+					NULL,
+					&error_local);
+	if (proxy == NULL) {
+		g_warning ("Failed to initialize d-bus proxy: %s",
+			   error_local->message);
+		return;
+	}
+	val = g_dbus_proxy_call_sync (proxy,
+					"ListGuards",
+					NULL,
+					G_DBUS_CALL_FLAGS_NONE,
+					-1,
+					NULL,
+					NULL);
+	data->bolt_supported = (val != NULL);
+	g_debug ("Bolt force power support: %d", data->bolt_supported);
+}
+
+static gboolean
+fu_plugin_thunderbolt_power_bolt_force_power (FuPlugin *plugin,
+					      GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_autoptr(GDBusConnection) connection = NULL;
+	g_autoptr(GDBusProxy) proxy = NULL;
+	g_autoptr(GUnixFDList) fds = NULL;
+	g_autoptr(GVariant) val = NULL;
+	GVariant *input;
+
+	input = g_variant_new ("(ss)",
+				"fwupd",   /* who */
+				"");       /* flags */
+
+	connection = g_bus_get_sync (G_BUS_TYPE_SYSTEM, NULL, error);
+	if (connection == NULL)
+		return FALSE;
+
+	proxy = g_dbus_proxy_new_sync (connection,
+					G_DBUS_PROXY_FLAGS_NONE,
+					NULL,
+					BOLT_DBUS_SERVICE,
+					BOLT_DBUS_PATH,
+					BOLT_DBUS_INTERFACE,
+					NULL,
+					error);
+	if (proxy == NULL)
+		return FALSE;
+
+	val = g_dbus_proxy_call_with_unix_fd_list_sync (proxy,
+							"ForcePower",
+							input,
+							G_DBUS_CALL_FLAGS_NONE,
+							-1,
+							NULL,
+							&fds,
+							NULL,
+							error);
+
+	if (val == NULL)
+		return FALSE;
+
+	if (g_unix_fd_list_get_length (fds) != 1) {
+		g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT,
+				"invalid number of file descriptors retunred: %d",
+				g_unix_fd_list_get_length (fds));
+		return FALSE;
+	}
+	data->bolt_fd = g_unix_fd_list_get (fds, 0, NULL);
```

Comment:
I'm confused why we need to use fd's like a handle -- why isn't bolt just using "NameOwnerChanged" to detect clients using ForcePower falling off the bus? The normal way clients do this kind of thing is with two methods, e.g. ForcePower() and UnForcePower() using a uint32_t "token" that allows a client to refcount. Falling off the bus is the same as UnForcePower()-ing for all senders.

Reply:
Now I am a bit confused :) I personally think that using an fd as "token" that can be easily released via `close` is more elegant than having an extra dbus method plus another uint token. On the bolt side the implementation also quite straight forward, create a named pipe with a `GIOChannel` as the even watch (and that is all for book-keeping). One cool thing about the named pipe is that we can recover from `boltd` crashes and restarts (`--replace`) because we just re-open all the named pipes that were not closed (and that works really well in my tests).
But maybe I am missing something, do you see any big disadvantage?